src/dbnode/persist/fs/cross_block_reader.go

```diff
@@ -24,33 +24,32 @@ import (
 	"bytes"
 	"container/heap"
 	"errors"
-	"fmt"
 	"io"
 	"time"
 
-	"github.com/m3db/m3/src/x/checked"
 	xerrors "github.com/m3db/m3/src/x/errors"
 	"github.com/m3db/m3/src/x/ident"
 	"github.com/m3db/m3/src/x/instrument"
-
-	"go.uber.org/zap"
 )
 
 var (
-	errReaderNotOrderedByIndex                = errors.New("crossBlockReader can only use DataFileSetReaders ordered by index")
-	errEmptyReader                            = errors.New("trying to read from empty reader")
-	_                          heap.Interface = (*minHeap)(nil)
+	errReaderNotOrderedByIndex     = errors.New("crossBlockReader can only use DataFileSetReaders ordered by index")
+	errUnorderedDataFileSetReaders = errors.New("dataFileSetReaders are not ordered by time")
+
+	_ heap.Interface = (*minHeap)(nil)
 )
 
 type crossBlockReader struct {
-	dataFileSetReaders []DataFileSetReader
-	id                 ident.ID
-	tags               ident.TagIterator
-	records            []BlockRecord
-	started            bool
-	minHeap            minHeap
-	err                error
-	iOpts              instrument.Options
+	dataFileSetReaders   []DataFileSetReader
+	pendingReaderIndices []int
```

Comment:
Any reason to use indices instead of just references to the readers themselves?

i.e.
```go
dataFileSetReaders   []DataFileSetReader
pendingDataFileSetReaders   []DataFileSetReader
```

Reply:
(not saying we should change it, just curious as to one vs other approach and whether you think there's advantages to one or the other, typically we've done the latter elsewhere so you can directly call methods on the references instead of then indexing into the other slice)

Reply:
I guess the only reason is that then I would have to put a `DataFileSetReader` into a `minHeapEntry` which I wanted to keep just a clean value object.

Reply:
Also, as you noticed, `minHeapEntry.dataFileSetReaderIndex` is used for ordering of the `minHeap` (to break the ties when the ids are equal), so I would need to have two values in order to follow your suggestion. Guess will leave this as is.